### PR TITLE
Set ViewStateUserKey For Classes Derived From Page

### PR DESCRIPTION
**Pull request by LingxiaChen**

### CODE REVIEW

File: src/Microsoft.NetCore.Analyzers/Core/Security/SetViewStateUserKey.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                {
+                    var classSymbol = (INamedTypeSymbol)symbolAnalysisContext.Symbol;
+                    var baseClassSymbol = classSymbol.BaseType;
+
+                    if (baseClassSymbol.Equals(pageTypeSymbol))
+                    {
+                        var onInitMethodSymbol = classSymbol.GetMembers().OfType<IMethodSymbol>().FirstOrDefault(s => s.Name == "OnInit" &&
+                                                                                                        s.Parameters.Length == 1 &&
+                                                                                                        s.Parameters[0].Type.Equals(eventArgsTypeSymbol) &&
+                                                                                                        s.IsProtected());
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Security/SetViewStateUserKey.cs]: General question: does the parameter here have to `eventArgsTypeSymbol` or any custom event args type derived from `eventArgsTypeSymbol`?

File: src/Microsoft.NetCore.Analyzers/Core/Security/SetViewStateUserKey.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+                compilationStartAnalysisContext.RegisterSymbolAction(symbolAnalysisContext =>
+                {
+                    var classSymbol = (INamedTypeSymbol)symbolAnalysisContext.Symbol;
+                    var baseClassSymbol = classSymbol.BaseType;
+
+                    if (baseClassSymbol.Equals(pageTypeSymbol))
+                    {
+                        var onInitMethodSymbol = classSymbol.GetMembers().OfType<IMethodSymbol>().FirstOrDefault(s => s.Name == "OnInit" &&
+                                                                                                        s.Parameters.Length == 1 &&
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Security/SetViewStateUserKey.cs]: Probably also want to check that method is non-static.

File: src/Microsoft.NetCore.Analyzers/UnitTests/Security/SetViewStateUserKeyTests.cs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        ViewStateUserKey = ""ViewStateUserKey"";
+    }
+}",
+            GetCSharpResultAt(5, 7, SetViewStateUserKey.Rule, "TestClass"));
+        }
+
+        [Fact]
+        public void TestSubclassWithSettingViewStateUserKeyNoDiagnostic()
+        {
+            VerifyCSharp(@"
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/UnitTests/Security/SetViewStateUserKeyTests.cs]: Will the following generate a false negative:

```
using System;
using System.Web.UI;

class MyType
{
    public string ViewStateUserKey { get; set; }
}

class TestClass : Page
{
    private MyType _field;
    protected override void OnInit (EventArgs e)
    {
        _field.ViewStateUserKey = ""ViewStateUserKey"";
    }
}
```

File: src/Microsoft.NetCore.Analyzers/Core/Security/SetViewStateUserKey.cs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                                                                                                        s.Parameters[0].Type.Equals(eventArgsTypeSymbol) &&
+                                                                                                        s.IsProtected());
+
+                        if (onInitMethodSymbol != null)
+                        {
+                            if (onInitMethodSymbol.GetTopmostOperationBlock(compilation)
+                                                    .Descendants()
+                                                    .Where(s => s is ISimpleAssignmentOperation simpleAssignmentOperation &&
+                                                                simpleAssignmentOperation.Target is IPropertyReferenceOperation propertyReferenceOperation &&
+                                                                propertyReferenceOperation.Property.Name == "ViewStateUserKey")
```
Comment (mavasani) [inline src/Microsoft.NetCore.Analyzers/Core/Security/SetViewStateUserKey.cs]: I would recommend additional checks to verify the type of the property matches your expected property type and the `IPropertyReferenceOperation.Instance` is non-null `IInstanceReferenceOperation` with `ReferenceKind` [InstanceReferenceKind.ContainingTypeInstance](http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/InstanceReferenceKind.cs,14)

File: src/Microsoft.NetCore.Analyzers/Core/Security/SetViewStateUserKey.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                }
+
+                compilationStartAnalysisContext.RegisterSymbolAction(symbolAnalysisContext =>
+                {
+                    var classSymbol = (INamedTypeSymbol)symbolAnalysisContext.Symbol;
+                    var baseClassSymbol = classSymbol.BaseType;
+
+                    if (baseClassSymbol.Equals(pageTypeSymbol))
+                    {
+                        var onInitMethodSymbol = classSymbol.GetMembers().OfType<IMethodSymbol>().FirstOrDefault(s => s.Name == "OnInit" &&
```
Comment (dotpaul) [inline src/Microsoft.NetCore.Analyzers/Core/Security/SetViewStateUserKey.cs]: What about the `Page_Init` method?

File: src/Microsoft.NetCore.Analyzers/Core/Security/SetViewStateUserKey.cs (hunk 1 above)
Comment (dotpaul) [reply src/Microsoft.NetCore.Analyzers/Core/Security/SetViewStateUserKey.cs]: Believe the only overload is EventArgs: https://docs.microsoft.com/en-us/dotnet/api/system.web.ui.page.oninit

File: src/Microsoft.NetCore.Analyzers/Core/Security/SetViewStateUserKey.cs (hunk 5 above)
Comment (LingxiaChen) [reply src/Microsoft.NetCore.Analyzers/Core/Security/SetViewStateUserKey.cs]: It seems there's no method named Page_Init but OnInit.

Comment (dotpaul) [reply src/Microsoft.NetCore.Analyzers/Core/Security/SetViewStateUserKey.cs]: Oh, I guess `Page_Init` is usually the method that's the event handler for the `Control.Init` (https://docs.microsoft.com/en-us/dotnet/api/system.web.ui.control.init) event.  It's not a virtual or abstract method.

I think we should look at it here too.

**Pull request merged**